src/server/session.ts

```diff
@@ -208,15 +208,29 @@ namespace ts.server {
             try {
                 if (this.operationHost.isCancellationRequested()) {
                     stop = true;
+                    tracing.instant(tracing.Phase.Session, "stepCancellation", { seq: this.requestId });
                 }
                 else {
-                    action(this);
+                    Debug.assert(!tracing.canPop());
+                    tracing.push(tracing.Phase.Session, "stepAction", { seq: this.requestId });
+                    try {
+                        action(this);
+                    }
+                    finally {
+                        while (tracing.canPop()) {
+                            tracing.pop();

```

Comment:
1. Re the `canPop` comment, related here: instead of adding that utility for this loop, just add some `popAll` method?  This could also be used with an exit handler to do the thing of dumping unterminated events in a case of `tsc` aborting for whatever reason.

2. And re the use of `try...catch` in other places: IIUC, it's to do the closing event in case of a cancel--?  If so, and given this code, wouldn't it be better to use `push...pop` which would all get dumped out by this bit?

3. Also, I'm probably confused, but given that this is already in a `try...catch`, the shouldn't this be done in the outer `catch`?

Reply:
1. I added `canPop` because I wanted to be able to consume it in asserts.  Once it existed, adding `popAll` seemed superfluous.  I have no objection to also adding an exit handler, though I don't actually know how to do that.
 2. Sorry, I don't understand your question.
 3. Based on some of the traces I've generated, it looks like an action can be cancelled without throwing (e.g. by detecting cancellation and exiting gracefully).  I used the try-finally to guard against this.

Reply:
Use `popAll`.

Reply:
(Iterate and then set length to 0).